### PR TITLE
[AST] Don't suggest unusable values for typo correction and code completion

### PR DESCRIPTION
**Pull request by rintaro**
Filter out:
- use within its own initial value
- use before declaration for local values

rdar://problem/25068938

### CODE REVIEW

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
Supersedes #20677.

@benlangmuir 
The logic shared with "use before declaration" diagnostics was [`FindLocalVal`](https://github.com/apple/swift/blob/master/lib/AST/NameLookupImpl.h#L35).
Since I implemented filter in `lookupVisibleDecls()`, we don't need to add any flag because all clients of it (code-completion, typo-correction, and refactoring) want "usable" values only.

File: lib/AST/LookupVisibleDecls.cpp [hunk 1]
```diff
@@ -1067,6 +1066,65 @@ void swift::lookupVisibleDecls(VisibleDeclConsumer &Consumer,
...
+    return;
+  }
+
+  // Filtering out unusable values.
+  class LocalConsumer : public VisibleDeclConsumer {
+    const SourceManager &SM;
+    SourceLoc Loc;
+    VisibleDeclConsumer &Consumer;
+
+    bool checkLocalValue(ValueDecl *VD, DeclVisibilityKind Reason) {
```
Comment (benlangmuir) [inline lib/AST/LookupVisibleDecls.cpp]: Please rename this to something that makes it clear what `true` and `false` return values mean - e.g. `isUsableLocalValue`.

File: lib/AST/LookupVisibleDecls.cpp [hunk 2]
```diff
@@ -1067,6 +1066,65 @@ void swift::lookupVisibleDecls(VisibleDeclConsumer &Consumer,
...
+      // Check "use within its own initial value" case.
+      if (auto *varD = dyn_cast<VarDecl>(VD))
+        if (auto *PBD = varD->getParentPatternBinding())
+          if (!PBD->isImplicit() &&
+              SM.rangeContainsTokenLoc(PBD->getSourceRange(), Loc))
+            return false;
+
+      switch (Reason) {
+      case DeclVisibilityKind::LocalVariable:
+        // Use 'TypeDecl's before declaration is allowed.
```
Comment (benlangmuir) [inline lib/AST/LookupVisibleDecls.cpp]: Nitpick: Use _of_

**Review by benlangmuir — APPROVED** [review]
Other than my comment about naming, LGTM.

**Comment by jrose-apple** [comment]
@DougGregor may also have opinions here from his work on ASTScope. (Specifically, lookupVisibleDecls isn't supposed to produce these invalid values in the first place.)

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by DougGregor — APPROVED** [review]
This is a good direction. It's what `ASTScope` would naturally do, but we don't want to wait for `ASTScope`.

**Pull request merged**